third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -859,12 +888,24 @@ void labelLoadOpsForTritonDot(scf::ForOp forOp) {
 
 struct PipelinePass : public TritonAMDGPUStreamPipelineBase<PipelinePass> {
   PipelinePass() = default;
-  PipelinePass(int32_t numStages, int32_t prefetch) {
+  PipelinePass(int32_t numStages, int32_t localStoreStage,
+               int32_t localLoadStage) {
     this->numStages = numStages;
-    this->prefetch = prefetch;
+
+    this->localStoreStage = localStoreStage;
+    this->localLoadStage = localLoadStage;
   }
 
   void runOnOperation() override {
+    // check numStages
+    if (localStoreStage < 0 ? numStages < -localStoreStage
+                            : localStoreStage >= numStages)
+      return;
+
+    if (localLoadStage < 0 ? numStages < -localLoadStage
+                           : localLoadStage >= numStages)
+      return;
```

Comment:
There is a hidden restriction/assumption of the current design, i.e. localStoreStage <= localLoadStage. 
However, the above conditions only checks localStoreStage or localLoadStage. E.g. if we have localStoreStage = 1, localLoadStage = -2 and num_stages = 2 they will pass the above conditions, but this is an illegal settings for a 2-stage pipeline.

Reply:
This check is just a preliminary sanity check against num_stages.

A more comprehensive check is handled in `StreamPipeliner::initSchedule` since it depends on maxIndirectionLevel.